openvdb/points/PointScatter.h

```diff
@@ -0,0 +1,676 @@
+        std::vector<Index64> values;
+        values.reserve(remainder);
+        for (Index64 i = 0; i < remainder; ++i) values.emplace_back(gen());
+        tbb::parallel_sort(values.begin(), values.end());
+
+        worker.join();
+
+        leafManager.foreach([&](typename BaseT::LeafNodeT& leaf, const size_t idx) {
+            const Index64 lowerOffset = voxelOffsets[idx]; // inclusive
+            const Index64 upperOffset = voxelOffsets[idx + 1]; // exclusive
+            assert(upperOffset > lowerOffset);
+
+            const auto valuesEnd = values.end();
+            auto lower = std::lower_bound(values.begin(), valuesEnd, lowerOffset);
+
+            auto* const data = leaf.buffer().data();
+            auto iter = leaf.beginValueOn();
+
+            Index32 currentOffset(0);
+            while (lower != valuesEnd) {
+                const Index64 vId = *lower;
+                if (vId >= upperOffset) break;
+
+                const Index32 nextOffset = Index32(vId - lowerOffset);
+                iter.increment(nextOffset - currentOffset);
+                currentOffset = nextOffset;
+                assert(iter);
+
+                auto& value = data[iter.pos()];
+                value = value + 1; // no += operator support
+                ++lower;
+            }
+        });
+
+        leafManager.foreach([&](typename BaseT::LeafNodeT& leaf, const size_t) {
+            if (BaseT::wasInterrupted()) return;
+            Index32 offset(0);
+            for (auto iter = leaf.beginValueAll(); iter; ++iter) {
+                if (iter.isValueOn()) {
+                    const Index32 value = mPointsPerVoxel + Index32(*iter);
+                    if (value == 0) leaf.setValueOff(iter.pos());
+                    else            offset += value;
+                }
+                // @note can't use iter.setValue(offset) on point grids
+                leaf.setOffsetOnly(iter.pos(), offset);
+            }
+
+            if (offset != 0) BaseT::generatePositions(leaf, offset);
+        }, /*threaded*/false);
+
+        if (mPointsPerVoxel == 0) tools::pruneInactive(BaseT::mPoints->tree());
+        BaseT::end();
+    }
+
+    /// @brief Print information about the scattered points
+    /// @parm name  A name to insert into the printed info
+    /// @parm os    The output stream
+    void print(const std::string &name, std::ostream& os = std::cout) const
+    {
+        if (!BaseT::mPoints) return;
+        const Index64 points = pointCount(BaseT::mPoints->tree());
+        const Index64 voxels = BaseT::mPoints->activeVoxelCount();
+
+        os << "Uniformly scattered " << points << " points into " << voxels
+           << " active voxels in \"" << name << "\" corresponding to "
+           << (double(points) / double(voxels)) << " points per voxel." << std::endl;
+    }
+
+private:
+    const Index64 mPointCount;
+    Index32 mPointsPerVoxel;
+};
+
+
+template <typename RandGenT,
```

Comment:
I'd recommend getting rid of this BasePointScatter inheritance for greater simplicity. In the case of tools/PointScatter.h, this base class stores all the point data and handles insertion, whereas here it's mostly used for storing some metadata and providing getters / setters.

Instead, I'd recommend wrapping the random number generator in a class that mutates the random values to fit the correct range and handles the spread, then pass this to be used as a member variable instead.

Reply:
On the other hand, it does provide some benefit. I think I'd personally turn the two generic functions "generatePositions" and "init" into static functions belonging to a point_scatter_internal namespace, but if you prefer the inheritance approach, that's ok with me.

Reply:
See comment on structure above